### PR TITLE
Scatter-Gather for Hexagon - Initial commit

### PR DESCRIPTION
**Pull request by aankit-ca**
@dsharletg @pranavb-ca @dpalermo @suyogsarda Please review the patch for scatter-gathers on hexagon.

### CODE REVIEW

File: apps/hexagon_histogram/pipeline.cpp [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+        RDom r(0, img_size);
+
+        Expr idx = clamp(input(r.x), 0, cast<uint16_t>(HIST_SIZE-1));
+        histogram(idx) += cast<HIST_TYPE>(1);
+        // Wrapper for the output
+        output(x) = histogram(x);
+
+        const int vector_size = get_target().has_feature(Target::HVX_128) ? 128 : 64;
+
+        histogram.vectorize(x, vector_size);
```
Comment (dsharletg) [inline apps/hexagon_histogram/pipeline.cpp]: Put all scheduling in one place or the other (not both generate() and schedule()).

File: apps/hexagon_histogram/process.cpp [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+    }
+    for (int x = 0; x < IMG_SIZE; x++) {
+        uint16_t idx = in_buf[x];
+        // Accumulator
+        calc_ptr[idx] += 1;
+    }
+    for (int x = 0; x < HIST_SIZE; x++) {
+        if (out_buf[x] != calc_ptr[x]) {
+            printf("Mismatch at %3d (x): %3lu (Halide) == %3lu (Expected)\n",
+                        x,
```
Comment (dsharletg) [inline apps/hexagon_histogram/process.cpp]: Indentation

File: src/HexagonOptimize.cpp [hunk 3]
```diff
@@ -1949,8 +1951,548 @@ class RearrangeExpressions : public IRMutator2 {
...
+                    // Make appropriate casts to handle 8-bit elements.
+                    if (old_elem_size == 1) {
+                        buffer_copy = cast(buffer_copy_ty.with_bits(16), buffer_copy);
+                        vtcm_copy = cast(vtcm_copy_ty.with_bits(8), vtcm_copy);
+                        value = cast(ty.with_bits(16), value);
+                    }
+                    Expr vtcm_base_var = Variable::make(Handle(), vtcm_buffer_name);
+                    // Place this call before the vscatter call.
+                    Stmt copy_to_vtcm = Store::make(vtcm_buffer_name, buffer_copy,
+                                              Ramp::make(buffer_at/new_elem_size, 1, const_extent),
```
Comment (dsharletg) [inline src/HexagonOptimize.cpp]: Indentation

File: src/HexagonOptimize.cpp [hunk 4]
```diff
@@ -1949,8 +1951,548 @@ class RearrangeExpressions : public IRMutator2 {
...
+                                              Parameter(), const_true(const_extent));
+                    // Place this call after the vscatter call.
+                    Stmt copy_from_vtcm = Store::make(op->name, vtcm_copy,
+                                                      Ramp::make(base, 1, const_extent),
+                                                      Parameter(),
+                                                      cast(op->predicate.type().with_lanes(const_extent),
+                                                           op->predicate));
+                    // Place this call before the copy from vtcm call
+                    Stmt scatter_sync = Evaluate::make(Call::make(Int(32),
+                                                   "halide_scatter_release",
```
Comment (dsharletg) [inline src/HexagonOptimize.cpp]: Indentation

File: src/HexagonOptimize.cpp [hunk 5]
```diff
@@ -1949,8 +1951,548 @@ class RearrangeExpressions : public IRMutator2 {
...
+                                                   Call::Extern));
+                    string suffix = (new_elem_size == 4) ? ".w.w" : ".h.h";
+                    if (is_acc) {
+                        suffix = "_add" + suffix;
+                    }
+
+                    Type out_ty = ty.with_bits(8 * new_elem_size);
+                    Expr shifted_index = simplify(cast(out_ty, index - base));
+                    stmt = Evaluate::make(Call::make(out_ty, "halide.hexagon.vscatter" + suffix,
+                                                   {vtcm_base_var, buffer_at, buff_mem - 1,
```
Comment (dsharletg) [inline src/HexagonOptimize.cpp]: Indentation

File: src/HexagonOptimize.cpp [hunk 6]
```diff
@@ -1949,8 +1951,548 @@ class RearrangeExpressions : public IRMutator2 {
...
+                    // If scatter is inside a parallel then we create 4 continuous slots.
+                    // Each thread acquires one slot for each iteration of the block.
+                    if (nested_par > 0) {
+                        Expr vtcm_slot_var = Variable::make(Int(32), "vtcm_slot" +
+                                                            std::to_string(nested_par));
+                        buffer_at = vtcm.alloc(op->name + ".buffer",
+                                               buff_mem * max_threads) +
+                                    buff_mem * vtcm_slot_var;
+                    }
+                    else {
```
Comment (dsharletg) [inline src/HexagonOptimize.cpp]: } else {

**Comment by steven-johnson** [comment]
Is this still under active development?

**Comment by aankit-ca** [comment]
Hi Steven,
I pushed a seperate PR and close this one https://github.com/halide/Halide/pull/3225
Thanks

**Pull request closed without merging**